base/util.jl

```diff
@@ -344,3 +344,85 @@ end
+        credbuf = Array(UInt8, 1024)
+        credbufsize = Ref{UInt32}(sizeof(credbuf))
+        succeeded = ccall((:CredPackAuthenticationBufferW, "credui.dll"), stdcall, Bool,
+            (UInt32, Cwstring, Cwstring, Ptr{UInt8}, Ptr{UInt32}),
+             CRED_PACK_GENERIC_CREDENTIALS, default_username, "", credbuf, credbufsize)
+        @assert succeeded
+
+        # Step 2: Create the actual dialog
+        #      2.1: Set up the window
+        messageArr = Base.cwstring(message)
+        captionArr = Base.cwstring(caption)
+        pfSave = Ref{Bool}(false)
+        cred = Ref{CREDUI_INFO}(CREDUI_INFO(sizeof(CREDUI_INFO), C_NULL, pointer(messageArr), pointer(captionArr), C_NULL))
+        dwflags = CREDUIWIN_GENERIC | CREDUIWIN_ENUMERATE_CURRENT_USER
+        if !prompt_username
+            # Disable setting anything other than default_username
+            dwflags |= CREDUIWIN_IN_CRED_ONLY
+        end
+        authPackage = Ref{Culong}(0)
+        outbuf_data = Ref{Ptr{Void}}(C_NULL)
+        outbuf_size = Ref{Culong}(0)
+
+        #      2.2: Do the actual request
+        code = ccall((:CredUIPromptForWindowsCredentialsW, "credui.dll"), stdcall, UInt32, (Ptr{CREDUI_INFO}, UInt32, Ptr{Culong},
+            Ptr{Void}, Culong, Ptr{Ptr{Void}}, Ptr{Culong}, Ptr{Bool}, UInt32), cred, 0, authPackage, credbuf, credbufsize[],
+            outbuf_data, outbuf_size, pfSave, dwflags)
+
+        #      2.3: If that failed for any reason other than the user canceling, error out.
+        #           If the user canceled, just return a nullable
+        if code == ERROR_CANCELLED
+            return Nullable{Tuple{String,String}}()
+        elseif code != ERROR_SUCCESS
+            error(Base.Libc.FormatMessage(code))
+        end
+
+        # Step 3: Convert encrypted credentials back to plain text
+        passbuf = Array(UInt16, 1024)
+        passlen = Ref{UInt32}(length(passbuf))
+        usernamebuf = Array(UInt16, 1024)
+        usernamelen = Ref{UInt32}(length(usernamebuf))
+        # Need valid buffers for domain, even though we don't care
+        dummybuf = Array(UInt16, 1024)
+        succeeded = ccall((:CredUnPackAuthenticationBufferW, "credui.dll"), Bool,
+            (UInt32, Ptr{Void}, UInt32, Ptr{UInt16}, Ptr{UInt32}, Ptr{UInt16}, Ptr{UInt32}, Ptr{UInt16}, Ptr{UInt32}),
+            0, outbuf_data[], outbuf_size[], usernamebuf, usernamelen, dummybuf, Ref{UInt32}(1024), passbuf, passlen)
+        if !succeeded
+            error(Base.Libc.FormatMessage())
+        end
+
+        # Step 4: Free the encrypted buffer
+        # ccall(:SecureZeroMemory, Ptr{Void}, (Ptr{Void}, Csize_t), outbuf_data[], outbuf_size[]) - not an actual function
+        ccall((:CoTaskMemFree, "ole32.dll"), Void, (Ptr{Void},), outbuf_data[])
+
+        # Done
+        Nullable((String(transcode(UInt8, usernamebuf[1:usernamelen[]-1])),
+            String(transcode(UInt8, passbuf[1:passlen[]-1]))))
```

Comment:
Potential security problems here — seems like multiple copies of the password could be left in memory: one for `passbuf`, one for `passbuf[1:passlen[]-1]`, and possibly `transcode` leaves partial copies as well (e.g. if it dynamically resizes a `UInt8` array).